src/vsr/journal.zig

```diff
@@ -1356,93 +1355,141 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
+                    const slot = Slot{ .index = index };
+
+                    // We only need to check the slot for op_prepare_max. Checked separately as
+                    // SlotRange.contains doesn't handle empty ranges.
+                    if (op_max == op_prepare_max - 1 and index != op_prepare_max_slot.index) {
+                        assert(op_max_to_op_prepare_max.head.index == op_prepare_max_slot.index);
+                        assert(op_max_to_op_prepare_max.tail.index == op_prepare_max_slot.index);
+                        continue;
+                    }
+
+                    if ((op_max == op_prepare_max - 1 and index == op_prepare_max_slot.index) or
+                        op_max_to_op_prepare_max.contains(slot))
+                    {
+                        const header_prepare_untrusted = &journal.headers[index];
+                        const header_prepare_ok = header_ok(
+                            replica.cluster,
+                            slot,
+                            header_prepare_untrusted,
+                        );
+                        const header_redundant_ok = header_ok(
+                            replica.cluster,
+                            slot,
+                            &journal.headers_redundant[index],
+                        );
+
+                        // We need our head op to be certain to reliably truncate torn prepares.
+                        // Head op is uncertain if we encounter one of the below faults:
+
+                        // 1. Corrupt redundant header or a misdirected read to a redundant header.
+                        if (header_redundant_ok == null) return torn_slots[0..0];
+
+                        // 2. Redundant header is set to .reserved. Could happen if:
+                        //    i. Slot was found corrupt on a previous startup, which set the header
+                        //       to reserved in memory.
+                        //    ii. Replica crashes *before* the corrupt slot was repaired, but
+                        //       *after* the reserved header was written to disk with a write to
+                        //       a nearby header (there are multiple headers in a single sector)
+                        if (header_redundant_ok.?.operation == .reserved) return torn_slots[0..0];
+
+                        // 3. Misdirected read for a prepare header.
+                        if (header_prepare_untrusted.valid_checksum()) {
+                            // This is a faulty slot and header is valid, prepare must be corrupt.
+                            assert(header_prepare_ok == null);
+                            return torn_slots[0..0];
+                        }
+
+                        // Header is valid and from a previous wrap.
+                        assert(header_redundant_ok != null);
+                        assert(header_redundant_ok.?.op < op_max);
```

Comment:
> Header is valid and from a previous wrap.

Can we also assert that the op is less than `op_checkpoint`?

Reply:
op wouldn't be strictly less than `op_checkpoint`; our range's tail is op_prepare_max's slot, which is the same as the current op_checkpoint's slot!

Added a <= assert in https://github.com/tigerbeetle/tigerbeetle/pull/2345/commits/8c0600bcefe1bab79a1df21d142fbe2c3d4a10d3.